File: src/lib/presets.ts
export type AspectPreset = {
  id: "16:9" | "9:16" | "1:1";
  label: string;
  description: string;
  ratio: [number, number];
};

export type SizePreset = {
  id: "720" | "1080" | "2160";
  label: string;
  description: string;
  longEdge: number;
};

export const aspectPresets: AspectPreset[] = [
  {
    id: "16:9",
    label: "16:9 Landscape",
    description: "YouTube, desktop",
    ratio: [16, 9]
  },
  {
    id: "9:16",
    label: "9:16 Vertical",
    description: "Shorts, Reels",
    ratio: [9, 16]
  },
  {
    id: "1:1",
    label: "1:1 Square",
    description: "Feeds, cover art",
    ratio: [1, 1]
  }
];

export const sizePresets: SizePreset[] = [
  {
    id: "720",
    label: "720p",
    description: "Fast export",
    longEdge: 1280
  },
  {
    id: "1080",
    label: "1080p",
    description: "Balanced",
    longEdge: 1920
  },
  {
    id: "2160",
    label: "4K",
    description: "Max quality",
    longEdge: 3840
  }
];

export function computeSize(aspect: AspectPreset, size: SizePreset) {
  const [w, h] = aspect.ratio;
  if (w >= h) {
    const height = Math.round((size.longEdge * h) / w);
    return { width: size.longEdge, height };
  }

  const width = Math.round((size.longEdge * w) / h);
  return { width, height: size.longEdge };
}

File: crates/core/src/presets.rs
pub struct AspectPreset {
  pub id: &'static str,
  pub label: &'static str,
  pub ratio: (u32, u32),
}

pub struct SizePreset {
  pub id: &'static str,
  pub label: &'static str,
  pub long_edge: u32,
}

pub const ASPECT_PRESETS: &[AspectPreset] = &[
  AspectPreset { id: "16:9", label: "16:9 Landscape", ratio: (16, 9) },
  AspectPreset { id: "9:16", label: "9:16 Vertical", ratio: (9, 16) },
  AspectPreset { id: "1:1", label: "1:1 Square", ratio: (1, 1) },
];

pub const SIZE_PRESETS: &[SizePreset] = &[
  SizePreset { id: "720", label: "720p", long_edge: 1280 },
  SizePreset { id: "1080", label: "1080p", long_edge: 1920 },
  SizePreset { id: "2160", label: "4K", long_edge: 3840 },
];

pub fn compute_size(aspect: &AspectPreset, size: &SizePreset) -> (u32, u32) {
  let (w, h) = aspect.ratio;
  if w >= h {
    let height = (size.long_edge * h) / w;
    (size.long_edge, height)
  } else {
    let width = (size.long_edge * w) / h;
    (width, size.long_edge)
  }
}
